### PR TITLE
feat(development): add PR review guidelines

### PR DESCRIPTION
**Pull request by sztok7**
# Description
Adds PR review guidelines for both reviewers and authors.

### CODE REVIEW

File: docs/3_development/1_development-flow/3_review.md [hunk 1]
```diff
@@ -29,4 +29,264 @@ If a project isn't setup with [codeowners](https://help.github.com/en/github/cre
 
 ## Changelog
 
-Assuming your project has releases, and is continually making updates, the project will need a changelog. When submitting a pr make sure to update the changelog reflecting the changes that your PR is making. This way, when a new release is made, there is already a changelog ready to go and no extra work is needed.
\ No newline at end of file
+Assuming your project has releases, and is continually making updates, the project will need a changelog. When submitting a pr make sure to update the changelog reflecting the changes that your PR is making. This way, when a new release is made, there is already a changelog ready to go and no extra work is needed.
```
Comment (mpetrunic) [inline docs/3_development/1_development-flow/3_review.md:32]: we should modify this, on most projects it's autogenerated

File: docs/3_development/1_development-flow/3_review.md [hunk 2]
```diff
@@ -29,4 +29,264 @@ If a project isn't setup with [codeowners](https://help.github.com/en/github/cre
...
+
+## Reviewer guide
+
+The guide for reviewers is mostly a condensed version of [Google’s code review best practices](https://google.github.io/eng-practices/review/reviewer/). Some parts are just c/p, some parts are emphasized, some things are left out, and some are added or are different.
+
+### The goal of the code review
+
+In general, reviewers should favor approving a PR once it is in a state where it definitely improves the overall code health of the system being worked on, even if the PR isn’t perfect.
```
Comment (mpetrunic) [inline docs/3_development/1_development-flow/3_review.md:40]: I would replace PR with "Code Change" everywhere

Comment (sztok7) [reply docs/3_development/1_development-flow/3_review.md]: https://github.com/ChainSafe/engineering-handbook/pull/107/commits/e85fb9323fcabd3ce964c1be910dd4aa9ed3d9f7

File: docs/3_development/1_development-flow/3_review.md [hunk 3]
```diff
@@ -11,13 +11,13 @@ Document everything, communicate with your team transparently, review together.
 
 - Ensure readability is retained in code e.g. complex nested ternary operators should have a comment note
 - Ensure adequate testing exists where needed
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:13]: I have a need to define `adequate` 🙈 but realized that is out of scope - maybe just link to QA docs?

File: docs/3_development/1_development-flow/3_review.md [hunk 4]
```diff
@@ -11,13 +11,13 @@ Document everything, communicate with your team transparently, review together.
...
-- PRs include a summary of work which outlines the objectives, decisions & considerations for the PR. Examples:
-    - This PR's objective was to implement middleware for logging, this was completed however presentation & storage of logs is incomplete pending discussion. Certain components are outside of the logging scope at the moment.
+- Code change requests include a summary of work which outlines the objectives, decisions & considerations for the code change. Examples:
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:14]: Do we want to link certain PR templates or at least mention as good to have?

File: docs/3_development/1_development-flow/3_review.md [hunk 5]
```diff
@@ -11,13 +11,13 @@ Document everything, communicate with your team transparently, review together.
...
+    - This PR's objective was to implement middleware for logging, this was completed however presentation & storage of logs is incomplete pending discussion. Certain components are outside the logging scope at the moment.
     - This PR includes minor styling tweaks on the desktop & mobile navigation menus, refactored navigation list logic.
 
 ## Communication
 
-Ask a team mate on the relevant team channel for a review, this ensures the request is clear as well as gives transparency to the rest of your team if commentary or considerations need to be made.
+Ask a teammate on the relevant team channel for a review, this ensures the request is clear as well as gives transparency to the rest of your team if commentary or considerations need to be made.
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:20]:
```suggestion
While platforms like Github allow selecting peer reviewers, additionally ask a teammate on the relevant team communication channel for a review. This ensures the request is clear and gives transparency to the rest of your team if commentary or considerations need to be made.
```

File: docs/3_development/1_development-flow/3_review.md [hunk 6]
```diff
@@ -11,13 +11,13 @@ Document everything, communicate with your team transparently, review together.
...
 
 ## Walkthroughs
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:22]: ```Walkthroughs are peer programming for reviews. If possible or required, ask for a walkthrough within your team. If your teammate has the capacity, reviewing together allows you to share the context of your work with your reviewer.```

File: docs/3_development/1_development-flow/3_review.md (hunk 5 above)
Comment (morrigan) [reply docs/3_development/1_development-flow/3_review.md:20]: Actually, makes sense that `Communication` part is under `Requesting reviews`

File: docs/3_development/1_development-flow/3_review.md [hunk 7]
```diff
@@ -27,6 +27,262 @@ Walkthroughs is peer programming for reviews, if possible/required, ask for a wa
 
 If a project isn't setup with [codeowners](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners) when you open a pull request, and mark it ready for review, you'll need to get your peers to review it. Every repository should have two different GitHub teams added to it (at minimum): `project_name-admins` and `project_name`, request the **non admin** team for review will allow your whole team to receive a notification about your new pull request.
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:28]: Suggesting structure changes:
```
If a project isn't setup with [codeowners](https://help.github.com/en/github/creating-cloning-and-archiving-repositories/about-code-owners) when you open a pull request, and mark it ready for review, you'll need to get your peers to review it. At a minimum, every repository should have two different GitHub teams added to it:
- `project_name-admins` - // define what it means
- `project_name` - // define what it means

Requesting the **non-admin** team for review will allow your whole team to receive a notification about your new pull request.
```

File: docs/3_development/1_development-flow/3_review.md [hunk 8]
```diff
@@ -27,6 +27,262 @@ Walkthroughs is peer programming for reviews, if possible/required, ask for a wa
...
 
-## Changelog
+## Reviewer guide
 
-Assuming your project has releases, and is continually making updates, the project will need a changelog. When submitting a pr make sure to update the changelog reflecting the changes that your PR is making. This way, when a new release is made, there is already a changelog ready to go and no extra work is needed.
\ No newline at end of file
+The guide for reviewers is mostly a condensed version of [Google’s code review best practices](https://google.github.io/eng-practices/review/reviewer/). Some parts are just c/p, some parts are emphasized, some things are left out, and some are added or are different.
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:32]:
```suggestion
The following guide for reviewers below is mostly a condensed and customized version of [Google’s code review best practices](https://google.github.io/eng-practices/review/reviewer/). For additional context we recommend reading the full guide but here are the most important points.
```

File: docs/3_development/1_development-flow/3_review.md [hunk 9]
```diff
@@ -27,6 +27,262 @@ Walkthroughs is peer programming for reviews, if possible/required, ask for a wa
...
+
+#### How fast should code reviews be
+
+If we are not in the middle of a focused task, we should do a code review shortly after it comes in.
+
+One business day is the maximum time it should take to respond to a code review request (i.e., the first thing the next morning).
+
+Following these guidelines means that a typical code change request should get multiple rounds of review (if needed) within a single day.
+
+📢 If we are in the middle of a focused task, such as writing code, we shouldn’t interrupt ourselves to do a code review. Instead, we should wait for a breakpoint in our work to respond to code change request.
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:117]:
```suggestion
:::note
If we are in the middle of a focused task, such as writing code, we shouldn’t interrupt ourselves to do a code review. Instead, we should wait for a breakpoint in our work to respond to code change request.
:::
```

File: docs/3_development/1_development-flow/3_review.md [hunk 10]
```diff
@@ -27,6 +27,262 @@ Walkthroughs is peer programming for reviews, if possible/required, ask for a wa
...
+
+#### Fast responses
+
+When we talk about fast code reviews, we are mainly talking about the response time. We should take time when reviewing the actual code change request. We need to fully understand the code. Ideally, the whole process should be fast. Remember, the code change doesn’t have to be perfect!
+
+📢 If we are too busy to do a full review, we can still send a quick response that lets the developer know when we will get to it. Or we can suggest another reviewer! We should delegate, and don’t allow the code change request to sour for days.
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:123]:
```suggestion
:::note
If we are too busy to do a full review, we can still send a quick response that lets the developer know when we will get to it. Or we can suggest another reviewer! We should delegate, and don’t allow the code change request to sour for days.
:::
```

File: docs/3_development/1_development-flow/3_review.md [hunk 11]
```diff
@@ -27,6 +27,262 @@ Walkthroughs is peer programming for reviews, if possible/required, ask for a wa
...
+
+Always:
+
+1. is it kind
+2. is it an absolute truth
+3. is it necessary
+
+#### Courtesy
+
+In general, it is important to be [courteous and respectful](https://chromium.googlesource.com/chromium/src/+/master/docs/cr_respect.md) while also being very clear and helpful to the developer whose code we are reviewing. One way to do this is to be sure that we are always making comments about the code and never making comments about the developer.
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:141]: maybe drop the link as it relates to Google culture

File: docs/3_development/1_development-flow/3_review.md [hunk 12]
```diff
@@ -27,6 +27,262 @@ Walkthroughs is peer programming for reviews, if possible/required, ask for a wa
...
+
+### References
+
+https://google.github.io/eng-practices/review/reviewer/
+
+https://phauer.com/2018/code-review-guidelines/#be-humble
+
+
+
+## Author guide
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:193]: Redunant, would remove

File: docs/3_development/1_development-flow/3_review.md [hunk 13]
```diff
@@ -27,6 +27,262 @@ Walkthroughs is peer programming for reviews, if possible/required, ask for a wa
...
+
+📢 We should **never respond in anger to code review comments.** That is a serious breach of professional etiquette that will live forever in the code review tool. Even if reviewers expressed their frustration first.
+
+#### We should fix the code before commenting
+
+If a reviewer says that they don’t understand something in our code, our first response should be to clarify the code itself! If the code can’t be clarified, we should add a code comment that explains why the code is there. If a comment seems pointless, only then should our response be an explanation in the code review tool.
+
+#### Authors can never ever resolve conversations
+
+Only the original reviewer can resolve the conversation.
```
Comment (morrigan) [inline docs/3_development/1_development-flow/3_review.md:271]:
```suggestion
Only the original reviewer who wrote the comment can resolve the conversation.
```

**Review by morrigan — CHANGES_REQUESTED** [review]
Thanks for putting this together! 🔥 

I have left smaller suggestions regarding certain structure and wording. Another point that comes to mind is recommending code formatting and style guidelines tools. They improve readability a lot so might be worth mentioning.

**Review by mpetrunic — APPROVED** [review]

File: docs/3_development/1_development-flow/3_review.md (hunk 11 above)
Comment (sztok7) [reply docs/3_development/1_development-flow/3_review.md:145]: but is it that tightly coupled with it? because it's just 8 DOs and 4 DON'Ts which seem pretty generic?

File: docs/3_development/1_development-flow/3_review.md (hunk 3 above)
Comment (sztok7) [reply docs/3_development/1_development-flow/3_review.md]: https://github.com/ChainSafe/engineering-handbook/pull/107/commits/4781b6d47a7e258410bf1a82d7dabe9bdf9e09e9

File: docs/3_development/1_development-flow/3_review.md (hunk 5 above)
Comment (sztok7) [reply docs/3_development/1_development-flow/3_review.md]: https://github.com/ChainSafe/engineering-handbook/pull/107/commits/2f85bd53bf75c909c33e7c68adb32a20f2d6f5c1

File: docs/3_development/1_development-flow/3_review.md (hunk 7 above)
Comment (sztok7) [reply docs/3_development/1_development-flow/3_review.md]: https://github.com/ChainSafe/engineering-handbook/pull/107/commits/2f85bd53bf75c909c33e7c68adb32a20f2d6f5c1

File: docs/3_development/1_development-flow/3_review.md (hunk 8 above)
Comment (sztok7) [reply docs/3_development/1_development-flow/3_review.md]: https://github.com/ChainSafe/engineering-handbook/pull/107/commits/514d590f3a27b4d7d1e74de5dcf488619e1b27a2

File: docs/3_development/1_development-flow/3_review.md (hunk 11 above)
Comment (mpetrunic) [reply docs/3_development/1_development-flow/3_review.md:134]: In any case it should.prpbably be permalink 🙂

Comment (sztok7) [reply docs/3_development/1_development-flow/3_review.md:134]: agreed, but i think this is the only way it's usable (they use it too) :sadge:

Comment (sztok7) [reply docs/3_development/1_development-flow/3_review.md:134]: i mean, i guess i can extract those dos and donts to this doc if we don't want to depend on Google haha

**Review by morrigan — APPROVED** [review]
👏
